### PR TITLE
refactor: extract user and item mention nodes via MDAST

### LINKED ISSUE #2589: Users are notified of mentions in code blocks
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

[This comment](https://stacker.news/items/1240888) mentions a few users.

They are not highlighted as a mention because it's in a code block (which is good) but notifications were still sent out. This is the case because detecting if somebody was mentioned on the server is not context aware:

https://github.com/stackernews/stacker.news/blob/288892e28b4bc8f2c179d04ef63516e64489d60b/api/paidAction/lib/item.js#L5-L22

### Screenshots

_No response_

### Steps To Reproduce

1. Mention someone in a code block
2. Users are notified

### Expected behavior

Users are not notified

### Logs

_No response_

### Device information

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Soxasora**
## Description

Fixes #2589 by extracting user and (also!) item mention nodes via MDAST.
`extractMentions` first parses the item's markdown into an MDAST tree with `itemMention` and `userMention` nodes support (via transforms), then walks the tree collecting them.

Also fixes path omissions e.g. `https://stacker.news/items/1/related -> #1` instead of `#1/related`.

bits:
- fix: `getDisplayText`, computes the display text for an item mention node. It returns custom text if available, otherwise parses the default label, with path, from the node's URL. If both fail, defaults to the item id.
  - extracted `parseItemUrl` from `parseInternalLinks` to avoid unnecessary checks for an already confirmed `ItemMentionNode`
- enhancement: item mentions are true MDAST nodes via `itemMentionTransform`, this made the MDAST walk trivial when extracting mentions
- cleanup: moved lexical utility from mdast shared utils

## Additional Context

It didn't make sense to use Lexical for this. The item create/update route doesn't have a Lexical state, it carries the `text` markdown ... so we would've needed MDAST conversion anyway.

## Checklist

**Are your changes backward compatible? Please answer below:**

Yes

**On a scale of 1-10 how well and how have you QA'd this change and any features it might affect? Please answer below:**

7, mention roundtrips, detection and notifications work correctly. also tested edge cases such as item mentions with image as `linkText`.

**Did you use AI for this? If so, how much did it assist you?**

Yes, review and e2e tests

---

> [!NOTE]
> **Medium Risk**
> Changes how @users and internal item links are detected for DB mentions and push notifications; behavior should align with markdown structure (code blocks, custom link text) but differs from the old plain-text regex approach.
> 
> **Overview**
> **Mention detection for item create/update** no longer scans raw text with regexes. A new server helper `extractMentions` parses markdown to MDAST, runs the same mention transforms as the editor, and collects `userMention` / `itemMention` nodes. `getMentions` and `getItemMentions` now take pre-extracted `names` and `itemIds` and only resolve them in the database.
> 
> **Item links in markdown** are promoted to first-class MDAST `itemMention` nodes via `itemMentionTransform` (before misleading-link rewriting), with a dedicated import visitor instead of inferring mentions from generic `link` nodes. Editor import/export and server extraction share this pipeline.
> 
> **Display labels** for item mentions use `getDisplayText()` / `parseItemUrl`, so URLs like `/items/1/related` show `#1/related` instead of dropping the path segment.
> 
> Small cleanups: `isImageOnlyLink` in mdast shared utils, `$isBlockElement` moved to Lexical node utils for list/quote export.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7c9470521cd8732a8c2dbc3bb341e11783700994. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: lib/lexical/server/mentions.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  mentionTransform(tree)
+  itemMentionTransform(tree)
+
+  const userNames = new Set()
+  const itemIds = new Set()
+
+  visit(tree, ['userMention', 'itemMention'], (node) => {
+    if (node.type === 'userMention') {
+      userNames.add(node.value.name)
```
[the remark this answers is not in the document]
Comment (Soxasora) [reply lib/lexical/server/mentions.js:35]: good ...?

File: lib/lexical/mdast/transforms/mentions.js [hunk 2]
```diff
@@ -36,3 +40,36 @@ export function mentionTransform (tree) {
...
+/** walk link nodes and replace internal item links with an itemMention node.
+*
+* must run before misleadingLinkTransform so we don't lose custom link text.
+*/
+export function itemMentionTransform (tree) {
+  visit(tree, 'link', (node, index, parent) => {
+    if (!node.url || !parent || index === undefined) return
+
+    // skip if link wraps only an image
+    if (isImageOnlyLink(node)) return
```
[the remark this answers is not in the document]
Comment (Soxasora) [reply lib/lexical/mdast/transforms/mentions.js:53]: good

**Comment by huumn** [comment]
looks good and works good!
